doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,153 @@
+
+There are a number of possibilities for how the long term QUIC API could be
+structured.
+
+The MVP will be written to allow current applications to use QUIC as a single
+stream at “a flip of a switch”.  Just enough of the chosen longer term
+design option is expected to be implemented _internally_ to support what the
+MVP needs (i.e. not necessarily published as a public API).
+
+
+## Candidate Solution: Just use the existing SSL API
+
+In this solution we just use the existing SSL API (e.g. with a
+“QUIC_client_method” and a “QUIC_server_method”) and extend it where
+necessary to support QUIC specific functionality. Existing functions (such as
+SSL_read or SSL_write) can be repurposed to work as you might expect (with a
+default stream) for QUIC.
+
+The existing SSL API does not separate the concepts of “connection” and
+“stream”. However in QUIC there can be multiple streams associated with a
+single connection. Therefore we would have to solve the problem of how to
+represent this.  There are several possibilities:
+
+* The `SSL` type represents a connection. We would need a new type for a stream
+* The `SSL` type represents a stream.  We would need a new type for a
+connection.
+* The `SSL` type represents a connection, and it has functions to target the
+desired stream, e.g.: \
+`SSL_read_stream(ssl, 4, buf, sizeof(buf));`
+* The `SSL` type represents both a connection and stream (in separate
+instances).  Stream objects can be created from a connection object as
+required, e.g.: \
+`SSL *SSL_create_stream(SSL *ssl, int stream_number, int type);`
+
+Pros
+
+* Re-use all of the existing APIs
+* Easier to convert existing applications to QUIC
+* The API is already understood by application authors - they don’t need to
+learn something new
+
+Cons
+
+* SSL objects are not currently thread safe. It’s not clear how the above
+model would work in the case where multiple streams associated with a single
+connection are being read from/written to in different threads
+* SSL objects have lots of APIs that are very specific to TLS and would not fit
+well with QUIC
+* We would have to “pollute” the SSL API with lots of QUIC specific
+functions
+* The “SSL” type name doesn’t fit well with QUIC
+
+
+## Candidate Solution: A generic comms API with SSL compat layer
+
+In this approach we would develop a completely new generic secure
+communications API that would support multiple protocol types, e.g. TLS, DLTS,
+QUIC, etc. We might expect to see separate types for generic “connections”
+and “streams”. Internally to a generic connection we would see protocol
+specific objects. We could choose to expose these internal protocol specific
+objects as part of the API, or we could expose them through generic mechanisms
+(e.g. ctrls, OSSL_PARAM etc).
+
+Some features can be simulated in protocols that don’t have them. For example
+TLS doesn’t have a multi-stream capability - but it could be simulated by
+running resumption handshakes to create a new “stream”.
```

Comment:
I think this part at least is straight forward. An attempt to create such a stream simply fails. We will have to handle this possibility anyway.

Reply:
In QUIC, it's not really possible to fail to create a stream, but all other streams continue to survive. A stream can be (temporarily) blocked on flow control. It can be created and then immediately shutdown (possibly in one or both directions; depending on the app behavior and if the stream is bidirectional or unidirectional) by the app, **with an application-specific error code**. I think it would be nearly impossible to imitate these behaviors exactly; resulting in an even more complicated application layer because it has to know which behavior to expect.